services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -0,0 +1,402 @@
+        }
+
+        String preAuthorizedCode = generateAuthorizationCodeForClientSession(clientSession);
+
+        CredentialsOffer theOffer = new CredentialsOffer()
+                .setCredentialIssuer(OID4VCIssuerWellKnownProvider.getIssuer(session.getContext()))
+                .setCredentialConfigurationIds(List.of(offeredCredential.getId()))
+                .setGrants(
+                        new PreAuthorizedGrant()
+                                .setPreAuthorizedCode(
+                                        new PreAuthorizedCode()
+                                                .setPreAuthorizedCode(preAuthorizedCode)
+                                                // not yet supported
+                                                .setUserPinRequired(false)));
+
+        LOGGER.debugf("Responding with offer: %s", theOffer);
+        return Response.ok()
+                .entity(theOffer)
+                .build();
+    }
+
+    /**
+     * Returns a verifiable credential
+     */
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path(CREDENTIAL_PATH)
+    public Response requestCredential(
+            CredentialRequest credentialRequestVO) {
+        LOGGER.debugf("Received credentials request %s.", credentialRequestVO);
+
+        // do first to fail fast on auth
+        UserSessionModel userSessionModel = getUserSessionModel();
+
+        Format requestedFormat = credentialRequestVO.getFormat();
+        String requestedCredential = credentialRequestVO.getCredentialIdentifier();
+
+        SupportedCredential supportedCredential = Optional
+                .ofNullable(OID4VCIssuerWellKnownProvider.getSupportedCredentials(this.session)
+                        .get(requestedCredential))
+                .orElseThrow(
+                        () -> {
+                            LOGGER.debugf("Unsupported credential %s was requested.", requestedCredential);
+                            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
+                        });
+
+        if (!supportedCredential.getFormat().equals(requestedFormat)) {
+            LOGGER.debugf("Format %s is not supported for credential %s.", requestedFormat, requestedCredential);
+            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_FORMAT));
+        }
+
+        CredentialResponse responseVO = new CredentialResponse();
+
+        Object theCredential = getCredential(userSessionModel, supportedCredential.getScope(), credentialRequestVO.getFormat());
+        switch (requestedFormat) {
+            case LDP_VC, JWT_VC, SD_JWT_VC -> responseVO.setCredential(theCredential);
+            default -> throw new BadRequestException(
+                    getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
+        }
+        return Response.ok().entity(responseVO)
```

Comment:
What about a fresh 'cNonce' and 'NonceExpiresIn' with the response?

This is needed to allow wallet to request again a new vc with the same access_token.

Reply:
As far as  I understand, cNonce is only used in proof-binding, e.g. in the proof-object provided by the requester. Since we don't support that in that PR, it would not help in any scenario. Or do I misunderstand something here?

Reply:
If proof is out of scope of this PR then there is no need to check proof (while processing the credential request) or to provide fresh values with the response or the error response. 

Keep in mind that for some formats that's not optional. For instance for `mso_mdoc` proof is a way to communicate the device key, that's mandatory. Also in SD-JWT-VC is highly desirable to pass a proof in order to assemble the `cnf` claim  in the issued sd-jwt to enable presentations using  key binding JWT.

Reply:
I agree, but the goal really was to not let the PR explode. Proof-binding should be added in the near future, thought.